identityserver.plugins.authenticationactions.redirect-action/src/main/java/io/curity/identityserver/plugin/RedirectAction/RedirectActionAuthenticationAction.java

```diff
@@ -0,0 +1,149 @@
+import se.curity.identityserver.sdk.authenticationaction.AuthenticationAction;
+import se.curity.identityserver.sdk.authenticationaction.AuthenticationActionResult;
+import se.curity.identityserver.sdk.errors.ErrorCode;
+import se.curity.identityserver.sdk.http.HttpResponse;
+import se.curity.identityserver.sdk.service.HttpClient;
+import se.curity.identityserver.sdk.service.Json;
+import se.curity.identityserver.sdk.service.SessionManager;
+import se.curity.identityserver.sdk.service.WebServiceClient;
+import se.curity.identityserver.sdk.service.WebServiceClientFactory;
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptor;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+
+import static se.curity.identityserver.sdk.authenticationaction.completions.RequiredActionCompletion.Redirect.redirect;
+
+public final class RedirectActionAuthenticationAction implements AuthenticationAction
+{
+    private final Logger _logger = LoggerFactory.getLogger(RedirectActionAuthenticationAction.class);
+    private final WebServiceClient _client;
+    private final Json _jsonService;
+    private final SessionManager _sessionManager;
+    private final String _redirectURI;
+
+    public RedirectActionAuthenticationAction(RedirectActionAuthenticationActionConfig configuration) throws URISyntaxException
+    {
+        _client = getWebServiceClient(configuration, configuration.getRedirectURL() + configuration.getResultPath());
+        _jsonService = configuration.getJsonService();
+        _sessionManager = configuration.getSessionManager();
+        _redirectURI = configuration.getRedirectURL() + configuration.getRedirectPath() + "?status=";
+    }
+
+    @Override
+    public AuthenticationActionResult apply(AuthenticationAttributes authenticationAttributes,
+                                            AuthenticatedSessions authenticatedSessions,
+                                            String authenticationTransactionId,
+                                            AuthenticatorDescriptor authenticatorDescriptor)
+    {
+        String key = authenticationAttributes.getSubject() + authenticatorDescriptor.getId();
+
+        Attribute parameterFromSession = _sessionManager.get(key);
+
+        if (parameterFromSession == null)
+        {
+            return redirectResponse(key);
+        }
+        else
+        {
+            HttpResponse response = getExternalStatus(parameterFromSession);
+
+            if (response.statusCode() != 200)
+            {
+                return AuthenticationActionResult.failedResult("Couldn't verify external status");
+            }
+
+            Map<String, Object> responseJson = response.body(HttpResponse.asJsonObject(_jsonService));
+
+            String externalUserId = (String) responseJson.get("externalUserId");
```

Comment:
Better not to cast but to check that the type is what's expected first

Reply:
All that's needed from this line forward is a `Comparable`. Perhaps the type checking and casting can be avoided all together?

Reply:
Sure you can: 
```java
Object obj = responseJson.get("externalUserId");

if (!(obj instanceof String)) {
    throw new InvalidStateException("Expected a String but didn't get one!")
}
String externalUserId = responseJson.get("externalUserId").toString();
```